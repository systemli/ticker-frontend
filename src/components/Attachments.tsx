import { FC } from 'react'
import {
    ButtonBack,
    ButtonNext,
    CarouselProvider,
    Dot,
    Image,
    Slide,
    Slider,
} from 'pure-react-carousel'
import styled from 'styled-components'
import { Attachment } from '../lib/types'
import 'pure-react-carousel/dist/react-carousel.es.css'
import { Button } from 'semantic-ui-react'

const Wrapper = styled(CarouselProvider)`
    position: relative;
`

const DotWrapper = styled.div`
    position: absolute;
    bottom: 5;
    width: 100%;
    display: block;
    text-align: center;
`

interface Props {
    attachments: Attachment[]
}

const Attachments: FC<Props> = props => {
    const renderButtonsAndDots = () => {
        // FIXME: This doesn't feel like a good pattern.
        // Tried to use styled components instead which didn't work right away.
        const style = {
            position: 'absolute',
            bottom: '50%',
            transform: 'translateY(50%)',
        }

        return (
            <>
                <DotWrapper>
                    <Button.Group size="mini">
                        {[...Array(props.attachments.length).keys()].map(
                            slide => (
                                <Button
                                    key={slide}
                                    as={Dot}
                                    icon="circle"
                                    slide={slide}
                                />
                            )
                        )}
                    </Button.Group>
                </DotWrapper>
                <Button
                    as={ButtonBack}
                    color="grey"
                    floated="left"
                    icon="arrow circle left"
                    size="large"
                    style={{ ...style, ...{ left: 5 } }}
                />
                <Button
                    as={ButtonNext}
                    color="grey"
                    floated="right"
                    icon="arrow circle right"
                    size="large"
                    style={{ ...style, ...{ right: 5 } }}
                />
            </>
        )
    }

    return (
        <Wrapper
            naturalSlideHeight={0.75}
            naturalSlideWidth={1}
            totalSlides={props.attachments.length}
        >
            <Slider>
                {props.attachments.map((image, key) => (
                    <Slide key={key} index={key}>
                        <Image
                            hasMasterSpinner={false}
                            src={image.url}
                            style={{ objectFit: 'scale-down' }}
                        />
                    </Slide>
                ))}
            </Slider>
            {props.attachments.length > 1 && renderButtonsAndDots()}
        </Wrapper>
    )
}

export default Attachments
