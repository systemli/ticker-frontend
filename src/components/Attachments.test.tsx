import { render } from '@testing-library/react'
import Attachments from './Attachments'

describe('Attachment', function () {
  test('renders single image correctly', function () {
    const attachments = [{ contentType: 'image/jpeg', url: 'https://example.com/image.jpg' }]
    const { asFragment } = render(<Attachments attachments={attachments} />)

    expect(asFragment()).toMatchSnapshot()
  })

  test('renders multiple images as slider', function () {
    const attachments = [
      { contentType: 'image/jpeg', url: 'https://example.com/image.jpg' },
      { contentType: 'image/jpeg', url: 'https://example.com/image.jpg' },
    ]
    const { asFragment } = render(<Attachments attachments={attachments} />)

    expect(asFragment()).toMatchSnapshot()
  })
})
