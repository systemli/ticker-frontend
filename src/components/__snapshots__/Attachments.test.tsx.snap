// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Attachment > renders multiple images as slider 1`] = `
<DocumentFragment>
  <div
    class="carousel sc-bczRLJ ehiAyy"
  >
    <div
      aria-label="slider"
      aria-live="polite"
      class="horizontalSlider___281Ls carousel__slider carousel__slider--horizontal"
      role="listbox"
      tabindex="0"
    >
      <div
        class="carousel__slider-tray-wrapper carousel__slider-tray-wrap--horizontal"
      >
        <div
          class="sliderTray___-vHFQ sliderAnimation___300FY carousel__slider-tray carousel__slider-tray--horizontal"
          style="width: 200%; transform: translateX(0%) translateX(0px); flex-direction: row;"
        >
          <div
            aria-label="slide"
            aria-selected="true"
            class="slide___3-Nqo slideHorizontal___1NzNV carousel__slide carousel__slide--visible"
            role="option"
            style="width: 50%; padding-bottom: 37.5%;"
            tabindex="0"
          >
            <div
              class="slideInner___2mfX9 carousel__inner-slide"
            >
              <div
                alt=""
                class="image___xtQGH carousel__image carousel__image--loading"
                orientation="horizontal"
                src="https://example.com/image.jpg"
              />
            </div>
          </div>
          <div
            aria-label="slide"
            aria-selected="false"
            class="slide___3-Nqo slideHorizontal___1NzNV carousel__slide carousel__slide--hidden"
            role="option"
            style="width: 50%; padding-bottom: 37.5%;"
            tabindex="-1"
          >
            <div
              class="slideInner___2mfX9 carousel__inner-slide"
            >
              <div
                alt=""
                class="image___xtQGH carousel__image carousel__image--loading"
                orientation="horizontal"
                src="https://example.com/image.jpg"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-gsnTZi UmfOC"
    >
      <div
        class="ui mini buttons"
      >
        <button
          aria-label="slide dot"
          class="dot___3c3SI carousel__dot carousel__dot--0 carousel__dot--selected ui icon button"
          disabled=""
          role="button"
        >
          <i
            aria-hidden="true"
            class="circle icon"
          />
        </button>
        <button
          aria-label="slide dot"
          class="dot___3c3SI carousel__dot carousel__dot--1 ui icon button"
          role="button"
        >
          <i
            aria-hidden="true"
            class="circle icon"
          />
        </button>
      </div>
    </div>
    <button
      aria-label="previous"
      class="buttonBack___1mlaL carousel__back-button ui grey large icon left floated button"
      disabled=""
      role="button"
      style="position: absolute; bottom: 50%; transform: translateY(50%); left: 5px;"
    >
      <i
        aria-hidden="true"
        class="arrow circle left icon"
      />
    </button>
    <button
      aria-label="next"
      class="buttonNext___2mOCa carousel__next-button ui grey large icon right floated button"
      role="button"
      style="position: absolute; bottom: 50%; transform: translateY(50%); right: 5px;"
    >
      <i
        aria-hidden="true"
        class="arrow circle right icon"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`Attachment > renders single image correctly 1`] = `
<DocumentFragment>
  <div
    class="carousel sc-bczRLJ ehiAyy"
  >
    <div
      aria-label="slider"
      aria-live="polite"
      class="horizontalSlider___281Ls carousel__slider carousel__slider--horizontal"
      role="listbox"
      tabindex="0"
    >
      <div
        class="carousel__slider-tray-wrapper carousel__slider-tray-wrap--horizontal"
      >
        <div
          class="sliderTray___-vHFQ sliderAnimation___300FY carousel__slider-tray carousel__slider-tray--horizontal"
          style="width: 100%; transform: translateX(0%) translateX(0px); flex-direction: row;"
        >
          <div
            aria-label="slide"
            aria-selected="true"
            class="slide___3-Nqo slideHorizontal___1NzNV carousel__slide carousel__slide--visible"
            role="option"
            style="width: 100%; padding-bottom: 75%;"
            tabindex="0"
          >
            <div
              class="slideInner___2mfX9 carousel__inner-slide"
            >
              <div
                alt=""
                class="image___xtQGH carousel__image carousel__image--loading"
                orientation="horizontal"
                src="https://example.com/image.jpg"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Attachment renders multiple images as slider 1`] = `
<DocumentFragment>
  <div
    class="carousel sc-bczRLJ ehiAyy"
  >
    <div
      aria-label="slider"
      aria-live="polite"
      class="horizontalSlider___281Ls carousel__slider carousel__slider--horizontal"
      role="listbox"
      tabindex="0"
    >
      <div
        class="carousel__slider-tray-wrapper carousel__slider-tray-wrap--horizontal"
      >
        <div
          class="sliderTray___-vHFQ sliderAnimation___300FY carousel__slider-tray carousel__slider-tray--horizontal"
          style="width: 200%; transform: translateX(0%) translateX(0px); flex-direction: row;"
        >
          <div
            aria-label="slide"
            aria-selected="true"
            class="slide___3-Nqo slideHorizontal___1NzNV carousel__slide carousel__slide--visible"
            role="option"
            style="width: 50%; padding-bottom: 37.5%;"
            tabindex="0"
          >
            <div
              class="slideInner___2mfX9 carousel__inner-slide"
            >
              <div
                alt=""
                class="image___xtQGH carousel__image carousel__image--loading"
                orientation="horizontal"
                src="https://example.com/image.jpg"
              />
            </div>
          </div>
          <div
            aria-label="slide"
            aria-selected="false"
            class="slide___3-Nqo slideHorizontal___1NzNV carousel__slide carousel__slide--hidden"
            role="option"
            style="width: 50%; padding-bottom: 37.5%;"
            tabindex="-1"
          >
            <div
              class="slideInner___2mfX9 carousel__inner-slide"
            >
              <div
                alt=""
                class="image___xtQGH carousel__image carousel__image--loading"
                orientation="horizontal"
                src="https://example.com/image.jpg"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-gsnTZi UmfOC"
    >
      <div
        class="ui mini buttons"
      >
        <button
          aria-label="slide dot"
          class="dot___3c3SI carousel__dot carousel__dot--0 carousel__dot--selected ui icon button"
          disabled=""
          role="button"
        >
          <i
            aria-hidden="true"
            class="circle icon"
          />
        </button>
        <button
          aria-label="slide dot"
          class="dot___3c3SI carousel__dot carousel__dot--1 ui icon button"
          role="button"
        >
          <i
            aria-hidden="true"
            class="circle icon"
          />
        </button>
      </div>
    </div>
    <button
      aria-label="previous"
      class="buttonBack___1mlaL carousel__back-button ui grey large icon left floated button"
      disabled=""
      role="button"
      style="position: absolute; bottom: 50%; transform: translateY(50%); left: 5px;"
    >
      <i
        aria-hidden="true"
        class="arrow circle left icon"
      />
    </button>
    <button
      aria-label="next"
      class="buttonNext___2mOCa carousel__next-button ui grey large icon right floated button"
      role="button"
      style="position: absolute; bottom: 50%; transform: translateY(50%); right: 5px;"
    >
      <i
        aria-hidden="true"
        class="arrow circle right icon"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`Attachment renders single image correctly 1`] = `
<DocumentFragment>
  <div
    class="carousel sc-bczRLJ ehiAyy"
  >
    <div
      aria-label="slider"
      aria-live="polite"
      class="horizontalSlider___281Ls carousel__slider carousel__slider--horizontal"
      role="listbox"
      tabindex="0"
    >
      <div
        class="carousel__slider-tray-wrapper carousel__slider-tray-wrap--horizontal"
      >
        <div
          class="sliderTray___-vHFQ sliderAnimation___300FY carousel__slider-tray carousel__slider-tray--horizontal"
          style="width: 100%; transform: translateX(0%) translateX(0px); flex-direction: row;"
        >
          <div
            aria-label="slide"
            aria-selected="true"
            class="slide___3-Nqo slideHorizontal___1NzNV carousel__slide carousel__slide--visible"
            role="option"
            style="width: 100%; padding-bottom: 75%;"
            tabindex="0"
          >
            <div
              class="slideInner___2mfX9 carousel__inner-slide"
            >
              <div
                alt=""
                class="image___xtQGH carousel__image carousel__image--loading"
                orientation="horizontal"
                src="https://example.com/image.jpg"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
